Extend the diagnostics plugin with uniqueness tests and JSON output.

michaelsproul opened the pull request:
I've been working on improving the diagnostic registration system so that it can:

* Check uniqueness of error codes *across the whole compiler*. The current method using `errorck.py` is prone to failure as it relies on simple text search - I found that it breaks when referencing an error's ident within a string (e.g. `"See also E0303"`).
* Provide JSON output of error metadata, to eventually facilitate HTML output, as well as tracking of which errors need descriptions. The current schema is:

```
<error code>: {
    "description": <long description>,
    "use_site": {
        "filename": <filename where error is used>,
        "line": <line in file where error is used>
    }
}
```

[Here's][metadata-dump] a pretty-printed sample dump for `librustc`.

One thing to note is that I had to move the diagnostics arrays out of the diagnostics modules. I really wanted to be able to capture error usage information, which only becomes available as a crate is compiled. Hence all invocations of `__build_diagnostics_array!` have been moved to the ends of their respective `lib.rs` files. I tried to avoid moving the array by making a plugin that expands to nothing but couldn't invoke it in item position and gave up on hackily generating a fake item. I also briefly considered using a lint, but it seemed like it would impossible to get access to the data stored in the thread-local storage.

The next step will be to generate a web page that lists each error with its rendered description and use site. Simple mapping and filtering of the metadata files also allows us to work out which error numbers are absent, which errors are unused and which need descriptions.

[metadata-dump]: https://gist.github.com/michaelsproul/3246846ff1bea71bd049

rust-highfive commented:
r? @nrc

(rust_highfive has picked a reviewer for you, use r? to override)

michaelsproul commented:
cc @brson

michaelsproul commented:
This is my first time using `stage0` conditional compilation, is there anything I need to do to mark those lines as removable in the next snapshot?

nrc commented:
@michaelsproul you don't need to do anything - some people leave comments, but it is not necessary

nrc commented on src/libsyntax/diagnostics/metadata.rs:
```diff
@@ -0,0 +1,134 @@
+pub use self::Uniqueness::*;
+
+// Directory to use for extended error JSON.
+const ERROR_METADATA_DIR: &'static str = "tmp/extended-errors";
```
It would be nice if the user could customise this path, perhaps through an environment variable.

nrc commented on src/libsyntax/diagnostics/metadata.rs:
```diff
@@ -0,0 +1,134 @@
+}
+
+/// Check that the errors used in a given crate are unique across the compiler.
+/// This is achieved by reading metadata files for other crates and checking them.
```
Nit: these comments would be better phrased in terms of the actual inputs to the functions, rather than there proposed uses. I.e., something like, "Checks that no errors defined in err_map are duplicated in any of the error metadata files in the directory given by crate_file". Since I don't think there is any requirement for anything here to be tied to crates.

nrc replied:
Sorry, should be ERROR_METADATA_DIR, not crate_file in the above comment.

Also, might be better to take the crate name as an arg, and find the corresponding file in these functions, rather than relying on the caller to do that.

michaelsproul replied:
Do you think I should go as far as renaming `crate_name` to something like `err_set_name`?

nrc commented:
Looks good, and I'm excited to have proper checking of error diagnostics across crates. One problem I see here is that we might compare metadata from an old build of a crate with a new build of the current crate. I think to avoid some false positives, that when a crate is built, we should delete its file and the files of any dependencies. Unfortunately, I don't know how to do that because rustc doesn't have access to the deps. I guess this could be built in to our build system, but that sounds kinda tough.

Anyway, if you've got any ideas for how to do this, it would be great, otherwise, I think this can land as is (well, with the other comments addressed).

Also, I think, if possible, it would be good to pull out all the diagnostics stuff into its own crate, rather than have it in libsyntax (obvs, doesn't need to be done to land this, but if you'd like to do it in the future it would be appreciated).

michaelsproul commented:
@nrc: I've addressed your comments as best I could. In light of lints possibly getting error codes (#24900) I completed removed all mentions of crates from the metadata module. I also made the metadata dir adjustable via the `ERROR_METADATA_DIR` environment variable.

As far as stale errors go, I think we're unlikely to run into issues unless a crate is renamed or an error is moved from one crate to another. In either case the error message about where the duplicate was found should allow the problem to be resolved quickly. The current code deals with old metadata for the same crate by ignoring any metadata with a matching name.

I'm also happy to pull the diagnostics stuff into its own crate at some point :smile:

michaelsproul commented:
Oh, I also marked some more stuff public, because I realised I'll need it for the HTML page generator.

nrc commented:
@bors r+

bors commented:
:pushpin: Commit 096d72a has been approved by `nrc`

<!-- @bors r=nrc 096d72ad1617d40279372649cf9396de04060a07 -->

bors commented:
:hourglass: Testing commit 096d72a with merge 79e8da7...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4733)

michaelsproul commented:
Whoops! I've fixed the test case and pushed a new commit.

alexcrichton commented:
@bors: r=nrc d27230b

bors commented:
:hourglass: Testing commit d27230b with merge ac5f595...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/4744), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/4751), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/4708), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/4732), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/4700), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/4766), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/4764), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4742), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/4435), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/4420), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/3232), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/3257)
